src/watertap_contrib/seto/unit_models/zero_order/chemical_softening_0D.py

```diff
@@ -0,0 +1,949 @@
+                )  # e-3
+
+        elif (
+            self.config.softening_procedure_type
+            is SofteningProcedureType.single_stage_lime_soda
+        ):
+
+            @self.Constraint(doc="Lime dosing")
+            def eq_CaOH_dosing(b):
+                return b.CaOH_dosing == pyunits.convert(
+                    (b.CO2_CaCO3 + b.Ca_hardness_CaCO3) * 0.56 * prop_in.flow_vol,
+                    to_units=pyunits.kg / pyunits.d,
+                )
+
+            @self.Constraint(doc="Soda dosing")
+            def eq_Na2CO3_dosing(b):
+                return b.Na2CO3_dosing == pyunits.convert(
+                    (b.Ca_hardness_nonCaCO3 + b.Mg_hardness_nonCaCO3)
+                    * 1.06
+                    * prop_in.flow_vol,
+                    to_units=pyunits.kg / pyunits.d,
+                )
+
+            @self.Constraint(doc="CO2 for first basin")
+            def eq_CO2_first_basin(b):
+                return (
+                    b.CO2_first_basin
+                    == pyunits.convert(
+                        (
+                            prop_in.conc_mass_comp["Alkalinity_2-"]
+                            + (b.Ca_hardness_nonCaCO3 + b.Mg_hardness_nonCaCO3)
+                            - b.Ca_CaCO3
+                            + prop_out.conc_mass_comp["Ca_2+"]
+                        )
+                        * b.Ca_CaCO3_conv
+                        * prop_in.flow_vol,
+                        to_units=pyunits.kg / pyunits.d,
+                    )
+                    * 0.44
+                )  # e-3
+
+        elif (
+            self.config.softening_procedure_type
+            is SofteningProcedureType.excess_lime_soda
+        ):
+
+            @self.Constraint(doc="Excess lime addition")
+            def eq_excess_CaOH(b):
+                return (
+                    b.excess_CaOH
+                    == (
+                        b.CO2_CaCO3
+                        + prop_in.conc_mass_comp["Alkalinity_2-"]
+                        + b.Mg_hardness_CaCO3
+                    )
+                    * 0.15
+                )
+
+            @self.Constraint(doc="Lime dosing")
+            def eq_CaOH_dosing(b):
+                return b.CaOH_dosing == pyunits.convert(
+                    (
+                        b.CO2_CaCO3
+                        + b.Ca_hardness_CaCO3
+                        + 2 * b.Mg_hardness_CaCO3
+                        + b.Mg_hardness_nonCaCO3
+                        + b.excess_CaOH
+                    )
+                    * 0.56
+                    * prop_in.flow_vol,
+                    to_units=pyunits.kg / pyunits.d,
+                )
+
+            def eq_Na2CO3_dosing(b):
+                return b.Na2CO3_dosing == pyunits.convert(
+                    (b.Ca_hardness_nonCaCO3 + b.Mg_hardness_nonCaCO3)
+                    * 1.06
+                    * prop_in.flow_vol,
+                    to_units=pyunits.kg / pyunits.d,
+                )
```

Comment:
delete- you add this as a constraint somewhere above

Reply:
This constraint is repeated for 2 softening configs that involve addition of Na2CO3.